idx/bigtable/bigtable.go

```diff
@@ -0,0 +1,550 @@
+				err = adminClient.CreateColumnFamily(ctx, tableName, COLUMN_FAMILY)
+				if err != nil {
+					log.Fatalf("btStore: failed to create cf %s/%s. %s", tableName, COLUMN_FAMILY, err)
+				}
+				err = adminClient.SetGCPolicy(ctx, tableName, COLUMN_FAMILY, bigtable.MaxVersionsPolicy(1))
+				if err != nil {
+					log.Fatalf("btStore: failed to set GCPolicy of %s/%s. %s", tableName, COLUMN_FAMILY, err)
+				}
+			} else if policy == "" {
+				log.Infof("bigtable-idx: column family %s/%s exists but has no GCPolicy. Creating it.", tableName, COLUMN_FAMILY)
+				err = adminClient.SetGCPolicy(ctx, tableName, COLUMN_FAMILY, bigtable.MaxVersionsPolicy(1))
+				if err != nil {
+					log.Fatalf("btStore: failed to set GCPolicy of %s/%s. %s", tableName, COLUMN_FAMILY, err)
+				}
+			}
+		}
+	}
+	client, err := bigtable.NewClient(ctx, gcpProject, bigtableInstance)
+	if err != nil {
+		log.Fatalf("failed to create bigtable client. %s", err)
+	}
+
+	idx := &BigtableIdx{
+		MemoryIdx: *memory.New(),
+		client:    client,
+		tbl:       client.Open(tableName),
+		shutdown:  make(chan struct{}),
+	}
+	if updateBigTableIdx {
+		idx.writeQueue = make(chan writeReq, writeQueueSize-writeMaxFlushSize)
+	}
+	updateInterval32 = uint32(updateInterval.Nanoseconds() / int64(time.Second))
+	return idx
+}
+
+func (b *BigtableIdx) Init() error {
+	log.Infof("initializing bigtable-idx. Project=%s, Instance=%s", gcpProject, bigtableInstance)
+	if err := b.MemoryIdx.Init(); err != nil {
+		return err
+	}
+	if updateBigTableIdx {
+		b.wg.Add(writeConcurrency)
+		for i := 0; i < writeConcurrency; i++ {
+			go b.processWriteQueue()
+		}
+		log.Infof("bigtable-idx: started %d writeQueue handlers", writeConcurrency)
+	}
+
+	b.rebuildIndex()
+	if maxStale > 0 {
+		if pruneInterval == 0 {
+			return fmt.Errorf("bigtable-idx: pruneInterval must be greater then 0")
+		}
+		b.wg.Add(1)
+		go b.prune()
+	}
+	return nil
+}
+
+func (b *BigtableIdx) Stop() {
+	b.MemoryIdx.Stop()
+	close(b.shutdown)
+	if updateBigTableIdx {
+		close(b.writeQueue)
+	}
+	b.wg.Wait()
+
+	err := b.client.Close()
+	if err != nil {
+		log.Errorf("bigtable-idx: Error closing bigtable client. %s", err)
+	}
+}
+
+func (b *BigtableIdx) Update(point schema.MetricPoint, partition int32) (idx.Archive, int32, bool) {
+	pre := time.Now()
+
+	archive, oldPartition, inMemory := b.MemoryIdx.Update(point, partition)
+
+	if !updateBigTableIdx {
+		statUpdateDuration.Value(time.Since(pre))
+		return archive, oldPartition, inMemory
+	}
+
+	if inMemory {
+		// bigtable uses partition ID in the key prefix, so an "update" that changes the partition for
+		// an existing metricDef will just create a new row in the table and wont remove the old row.
+		// So we need to explicitly delete the old entry.
+		if oldPartition != partition {
+			go func() {
+				for {
+					err := b.deleteRow(FormatRowKey(archive.Id, oldPartition))
+					if err != nil {
+						log.Errorf("bigtable-idx: failed to delete row. %s", err)
```

Comment:
if a problem is recoverable, i think it should be a warning.
if i see this error logged, i would assume the row was never deleted, which may not be true here.

Reply:
https://godoc.org/cloud.google.com/go/bigtable#hdr-Retries

If there is a transient error (ie recoverable), then the delete will automatically be retried.